### PR TITLE
remove ResetValues since we are using reuseValues

### PR DESCRIPTION
**Pull request by andriisoldatenko**
from helm docs:
```
--reset-values                 when upgrading, reset the values to the ones built into the chart
--reuse-values                 when upgrading, reuse the last release's values and merge in any overrides from the command line via --set and -f. If '--reset-values' is specified, this is ignored
```

### CODE REVIEW

**Comment by sjmiller609** [comment]
I assumed invalidly that --reuse-values and --reset-values work together. the reason the test has failed now is with only --reuse-values, helm upgrade fails to upgrade airflow-chart from 0.11.1 to 0.15.0-alpha.1 with the nil pointer error.

The nil pointer error is a helm gotcha in both helm2 and helm3 where if you upgrade a chart, you need to provide a user-provided configuration for all new values in values.yaml.

The behavior we want is values to be selected in the following priority:

1. user defined value with -f config.yaml
2. existing value, from --reuse-values
3. default value

Andrii has discovered today that with --reset-values only, we are ignoring step 2. This is the bugfix being addressed by this PR.

The current behavior in prod cloud is to ignore step 3 (nil pointer error) and workaround by ensuring priority 3 is never encountered - all values provided in either 1 or 2.

Andrii and I discussed in PM, and the following option can accomplish what we want:

- --reset-values ON
- --reuse-values OFF
- -f config from houston
- -f full, existing config

where existing config is generated with:

helm3 get values --all

this provides the complete values including all computed values. This is basically a way to rig it up to get both --reset-values benefit (helm 3 only) and the same functionality as --reuse-values (using our special logic)

**Comment by andriisoldatenko** [comment]
> one thing to note is that this plan does not allow us to change cloud's default configurations by changing the default value in values.yaml. the longer term fix is to use --reset-values and provide all user-provided values every time, rendering the --reuse-values behavior pointless. This is the cleanest option, but it requires platform change to store all user-provided values in astro db.

I think it's doable same way i did in helm 2 https://github.com/astronomer/commander/pull/22/files

**Comment by sjmiller609** [comment]
> > one thing to note is that this plan does not allow us to change cloud's default configurations by changing the default value in values.yaml. the longer term fix is to use --reset-values and provide all user-provided values every time, rendering the --reuse-values behavior pointless. This is the cleanest option, but it requires platform change to store all user-provided values in astro db.
> 
> I think it's doable same way i did in helm 2 https://github.com/astronomer/commander/pull/22/files

that does not work to accomplish the behavior of changing an existing default in airflow-chart/values.yaml, because in this model all values get set as user-provided values, so they will overwrite the --reset-values default with user-provided

**Comment by schnie** [comment]
Well all those secrets that are generated once at creation need to be carried forward. These are all passwords so they are generated, hashed and stored in kubernetes secret. Houston only needs to know about the hahsed values of some of these since its doing pw validation.

What would we do about these values?

**Comment by sjmiller609** [comment]
> Well all those secrets that are generated once at creation need to be carried forward. These are all passwords so they are generated, hashed and stored in kubernetes secret. Houston only needs to know about the hahsed values of some of these since its doing pw validation.
> 
> What would we do about these values?

the existing plan:
--reset-values ON
--reuse-values OFF
-f config from houston
-f full, existing config

step 4 addresses those values.

**Comment by schnie** [comment]
Yea but how do we get those values? Do we `helm get values` before we deploy?

**Comment by andriisoldatenko** [comment]
@sjmiller609 you are reading my old-PR incorrectly: here is lline:
```
currentConfig, err = chartutil.ReadValues([]byte(r.Release.Config.Raw))
```

**Comment by sjmiller609** [comment]
the only situation NOT addressed by that plan changing the default value of an existing, non-user-provided value in airflow chart, and expecting that to update airflow releases in cloud. For example, if we wanted to reconfigure the default replicas on webserver https://github.com/astronomer/airflow-chart/blob/a73bcd6bff87d3c4eac37f8dd7036a06a91a0402/values.yaml#L228 and switch on this line to "2", then cloud would remain at "1" until we add into additional user-provided values by configuring houston. The best behavior would be to automatically reset to 2 if we never provided that value from houston.

**Comment by andriisoldatenko** [comment]
@schnie 
```
// before deleting we need to reuse values
		// reminder: revision = 0, means get Last release from pkg/tiller/release_content.go:33
		r, err := c.helm.ReleaseContent(releaseName, helm.ContentReleaseVersion(0))
		if err != nil {
			return "", err
		}
```

**Comment by sjmiller609** [comment]
> Yea but how do we get those values? Do we `helm get values` before we deploy?

yes, that's what I said in my first comment

**Comment by sjmiller609** [comment]
> @sjmiller609 you are reading my old-PR incorrectly: here is lline:
> 
> ```
> currentConfig, err = chartutil.ReadValues([]byte(r.Release.Config.Raw))
> ```
> 
> that's how you can read current/exisiting values(secrets etc from last release) and merge with new values and than run ResetValues using mergedValues.

very good, yes I agree this can work to solve the secondary issue I'm talking about!

**Comment by andriisoldatenko** [comment]
@sjmiller609 i think you are trying to fix all issues in one ticket.

**Comment by sjmiller609** [comment]
> @sjmiller609 i think you are trying to fix all issues in one ticket.

no I am not, I already mentioned that I consider the second issue out of scope

**Comment by schnie** [comment]
Soo.. do we merge this?

**Comment by andriisoldatenko** [comment]
@sjmiller609 sorry don't understand where is issue 1/ or 2/ and what to fix in this specific PR?

**Comment by sjmiller609** [comment]
we should not merge because of nil pointer issue, instead we should follow this plan:

--reset-values ON
--reuse-values OFF
-f config from houston
-f full, existing config

**Comment by schnie** [comment]
So is commander already grabbing the _existing_ configuration or do we have to add code to do that?

File: helm/client.go [hunk 1]
```diff
@@ -174,14 +174,42 @@ func (c *Client) UpdateRelease(releaseName, chartName, chartVersion, namespace s
...
+		get_values_client := action.NewGetValues(actionConfig)
+		get_values_client.AllValues = true
+		existing_values, err := get_values_client.Run(releaseName)
+		if err != nil {
+			logger.Errorf("Failed to run 'helm get values -n %s --all %s': %v", namespace, releaseName, err)
+			return nil, err
+		}
+		logger.Infof("Merging existing helm values with provided values of release %s", releaseName)
+		// Gotcha: you must re-load the values because they were mutated by the call to Run
+		ch, err := loader.Load(chartPath)
```
Comment (sjmiller609) [inline helm/client.go:190]: @andriisoldatenko I think this was the gotcha that was making yours have an issue

File: helm/client.go [hunk 2]
```diff
@@ -174,14 +174,42 @@ func (c *Client) UpdateRelease(releaseName, chartName, chartVersion, namespace s
...
 		// force resource updates through a replacement strategy
-		client.Force = true
-		logger.Infof("Retrying with --force flag to update %s", releaseName)
-		rel, err := client.Run(releaseName, ch, options)
+		// TODO: turn force back on after this issue is resolved
+		// https://github.com/kubernetes/kubernetes/issues/91459
+		// --force does not work for now because 'kubectl replace' is used
+		// under-the-hood, and this has a bug that a kube service can't be
+		// replaced without an immutable error, even when there is no change.
+		client.Force = false
```
Comment (sjmiller609) [inline helm/client.go:202]: There is an issue with --force command right now in helm, but it's a kube bug, I think https://github.com/kubernetes/kubernetes/issues/91459

File: kubernetes/namespace.go [hunk 3]
```diff
@@ -13,7 +14,8 @@ type Namespace struct {
 }
 
 func (n *Namespace) Exists(namespace string) (bool, error) {
-	_, err := n.ClientSet.CoreV1().Namespaces().Get(namespace, metav1.GetOptions{})
+	// https://godoc.org/context#TODO
+	_, err := n.ClientSet.CoreV1().Namespaces().Get(context.TODO(), namespace, metav1.GetOptions{})
```
Comment (sjmiller609) [inline kubernetes/namespace.go:18]: "context" is a go thing for managing interrupts and stuff. TODO means you did not decide how to manage it yet. After upgrading to helm 3.2.1, I was required to add context to the kube calls.

File: helm/client.go (hunk 1 above)
Comment (sjmiller609) [reply helm/client.go:190]: I found that the ch.Values was missing values _only after_ the call to client.Run

File: helm/client.go (hunk 2 above)
Comment (sjmiller609) [reply helm/client.go:202]: If there is a situation that could cause an immutable error, then let's add a test case for that.

**Comment by sjmiller609** [comment]
I ended up doing a lot of various things to get this working, but some of them did not matter. I left some in that we need anyways.

what I added:
- upgrade helm to 3.2.1 (not required, but might as well leave it in)
- delete testing namespaces when tests pass (to avoid overloading resources)
- add a test to cover the situation of a user-provided config being reset (to cover the issue currently on master branch due to my change)
- fix this test

notable:
- --force is not used, and also no delete and re-install fallback. I think we should add --force back after the issues is resolved, but in the meantime we should know ahead of time before there will be an immutable change.

File: helm/client.go (hunk 1 above)
Comment (sjmiller609) [reply helm/client.go:190]: this cost me like 2 hours, lol

File: test/component/features/update_deployment.feature [hunk 4]
```diff
@@ -1,25 +1,39 @@
 Feature: Update Deployment
 
-  Scenario: Update Airflow chart version
+  Scenario: Update Airflow chart version with user provided values during create
      Given a kubernetes client
       And a namespace astronomer-sample-upgrade
       And platform release name is astronomer
-      When we call CreateDeployment sample-upgrade into namespace astronomer-sample-upgrade with airflow-chart version 0.11.1
+      When we call CreateDeployment sample-upgrade into namespace astronomer-sample-upgrade with airflow-chart version 0.11.1 with --set fernetKey=TGlEeVBoU3RUTVk1TEk5b0VwYUtiZHViZjNOS2dRRWo=
```
Comment (sjmiller609) [inline test/component/features/update_deployment.feature:7]: this is the test case that is failing on master. my existing test was not covering the situation that I thought it was.

**Pull request merged**